### PR TITLE
new L1 <-> L2 msg passing

### PR DESCRIPTION
**Pull request by axchu**
upgrade to hardhat included

### CODE REVIEW

File: contracts/ERC20/L1ERC20Deposit.sol [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            _amount
+        );
+        
+        // Generate encoded calldata to be executed on L2.
+        bytes memory message = abi.encodeWithSignature(
+            "deposit(address,uint256)",
+            _depositer,
+            _amount
+        );
+        messenger.sendMessage(l2ERC20Address, message, 1000000);
```
Comment (karlfloersch) [inline contracts/ERC20/L1ERC20Deposit.sol:38]: Long term might be worth metering this and coming up with a bit more of a lower bounded value. That said looks good to me for now!

File: contracts/ERC20/L1ERC20Deposit.sol [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    function withdraw( 
+        address _withdrawer,
+        uint256 _amount
+    ) public {
+        require(l2ERC20Address == messenger.xDomainMessageSender());
+        l1ERC20.transfer(_withdrawer, _amount); 
+    }
+
```
Comment (karlfloersch) [inline contracts/ERC20/L1ERC20Deposit.sol:48]: extra whitespace!

File: contracts/ERC20/L2ERC20.sol [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+contract L2ERC20 is ERC20 {
+    address l1ERC20DepositAddress;
+    iOVM_BaseCrossDomainMessenger internal messenger;
+
+    constructor(
+        uint256 _initialAmount,
+        string memory _tokenName,
+        uint8 _decimalUnits,
+        string memory _tokenSymbol
+    ) public ERC20(_initialAmount, _tokenName, _decimalUnits, _tokenSymbol) {}
```
Comment (karlfloersch) [inline contracts/ERC20/L2ERC20.sol:17]: Ah probably makes sense to hardcode `_initialAmount` to zero

File: contracts/ERC20/L2ERC20.sol [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    function init(
+        address _messenger,
+        address _L1ERC20DepositAddress
+    ) public {
+        require(l1ERC20DepositAddress == address(0), "L2ERC20 instance has already been initalized");
+        messenger = iOVM_BaseCrossDomainMessenger(_messenger);
+        l1ERC20DepositAddress = _L1ERC20DepositAddress;
+    }
```
Comment (karlfloersch) [inline contracts/ERC20/L2ERC20.sol:26]: Can we put the `init` function in the constructor?

File: contracts/ERC20/L2ERC20.sol [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    function deposit(address _depositor, uint256 _amount) public returns (bool success) {
```
Comment (karlfloersch) [inline contracts/ERC20/L2ERC20.sol:28]: I'd probably name this `mint(...)` to be closer to ERC20 conventions, though could be convinced otherwise. Specially reading it as something like `mint(...){require(thisIsCalledByL1Messenger); ...}` you know it's minting & only callable by L1.

File: contracts/ERC20/L2ERC20.sol [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        require(messenger.xDomainMessageSender() == l1ERC20DepositAddress); //??
```
Comment (karlfloersch) [inline contracts/ERC20/L2ERC20.sol:29]: This looks right! Can get rid of `//??`

File: test/erc20/simplified-message-passing.spec.ts [hunk 7]
```diff
@@ -1,25 +1,30 @@
...
+import { assert } from 'console'
 
 /* Internal Imports */
-import { increaseEthTime } from '../helpers'
+//import { increaseEthTime } from '../helpers'
 
-const l1ToL2MessageDelay = 100
-const l2ToL1MessageDelay = 1000
+const l1ToL2MessageDelay = 0 //5 * 60 //5 minutes
+const l2ToL1MessageDelay = 0 //60 * 60 * 24 * 7 //1 week
```
Comment (karlfloersch) [inline test/erc20/simplified-message-passing.spec.ts:13]: Is this supposed to be zero'd out still? Is this because the simulated timeout stuff doesn't work at the moment? Should we add a `// TODO` ?

File: test/erc20/simplified-message-passing.spec.ts [hunk 8]
```diff
@@ -1,25 +1,30 @@
...
-describe('L1 <=> L2 ERC20 (Simplified Example)', () => {
-  let signer: Signer
+describe('EOA L1 <-> L2 Message Passing', () => {
+  let AliceL1Wallet: Signer
+  let BobL1Wallet: Signer
+  let MalloryL1Wallet: Signer
   before(async () => {
-    ;[signer] = await ethers.getSigners()
+    // TODO: Update this to attach to different ethers providers within buidler
+    // as to properly simulate cross domain calls.
```
Comment (karlfloersch) [inline test/erc20/simplified-message-passing.spec.ts:21]: Is this related to the timeout stuff above?

File: test/erc20/simplified-message-passing.spec.ts [hunk 9]
```diff
@@ -36,103 +41,129 @@ describe('L1 <=> L2 ERC20 (Simplified Example)', () => {
...
+      await L1ERC20Deposit.deposit(AliceL1Wallet.getAddress(), 5000)
+      await relayL1ToL2Messages(signer)  
 
-      expect(finalL1Balance).to.equal(0)
-      expect(finalL2Balance).to.equal(originalL1Balance.add(originalL2Balance))
+      await expect(L2ERC20.connect(MalloryL1Wallet).withdraw(3000)).to.be.revertedWith("account doesn't have enough coins to burn")
     })
+
+
+
```
Comment (karlfloersch) [inline test/erc20/simplified-message-passing.spec.ts:167]: Whitespace!

**Review by karlfloersch — APPROVED** [review]
Awesome!!! This looks great! Huge fan of this!

I left a few suuuper minor stylistic nitpicks for the most part. The only tangible thing is I also just opened a quick PR which adds my (working) `yarn.lock` -- https://github.com/ethereum-optimism/Buidler-Deposit-Withdraw-Example/pull/9

File: test/erc20/simplified-message-passing.spec.ts (hunk 7 above)
Comment (axchu) [reply test/erc20/simplified-message-passing.spec.ts:13]: added TODOs

File: test/erc20/simplified-message-passing.spec.ts (hunk 8 above)
Comment (axchu) [reply test/erc20/simplified-message-passing.spec.ts:21]: it's a legacy comment from earlier versions...Ben says ok to remove

File: contracts/ERC20/L2ERC20.sol [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        address _messenger,
+        address _L1ERC20DepositAddress
+    ) public {
+        require(l1ERC20DepositAddress == address(0), "L2ERC20 instance has already been initalized");
+        messenger = iOVM_BaseCrossDomainMessenger(_messenger);
+        l1ERC20DepositAddress = _L1ERC20DepositAddress;
+    }
+
+    function mint(address _depositor, uint256 _amount) public returns (bool success) {
+        require(messenger.xDomainMessageSender() == l1ERC20DepositAddress);
```
Comment (K-Ho) [inline contracts/ERC20/L2ERC20.sol:29]: add check that `msg.sender` is the L2CrossDomainMessenger
```suggestion
        require(messenger.xDomainMessageSender() == l1ERC20DepositAddress);
        require(msg.sender == address(messenger), "Only messages relayed by L2CrossDomainMessenger can mint")
```

File: contracts/ERC20/L1ERC20Deposit.sol [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+            _amount
+        );
+        messenger.sendMessage(l2ERC20Address, message, 1000000); //TODO: meter this, find a lower-bounded value
+    }
+
+    function withdraw(
+        address _withdrawer,
+        uint256 _amount
+    ) public {
+        require(l2ERC20Address == messenger.xDomainMessageSender());
```
Comment (K-Ho) [inline contracts/ERC20/L1ERC20Deposit.sol:45]: add check that msg.sender is L1CrossDomainMessenger
```suggestion
        require(l2ERC20Address == messenger.xDomainMessageSender());
        require(msg.sender == address(messenger), "Only messages relayed by the L1CrossDomainMessenger can withdraw")
```

File: contracts/ERC20/L2ERC20.sol (hunk 3 above)
Comment (K-Ho) [inline contracts/ERC20/L2ERC20.sol:17]: +1 to Karl's comment!
```suggestion
    constructor(
        string memory _tokenName,
        uint8 _decimalUnits,
        string memory _tokenSymbol
    ) public ERC20(0, _tokenName, _decimalUnits, _tokenSymbol) {}
```

Comment (K-Ho) [reply contracts/ERC20/L2ERC20.sol:17]: This is because the L2 ERC20 counterpart to an L1 native ERC20 will start off with 0 supply

File: contracts/ERC20/L2ERC20.sol (hunk 4 above)
Comment (K-Ho) [reply contracts/ERC20/L2ERC20.sol:26]: I believe we need this - the deposit contract takes in the L2ERC20 address as constructor param, so the L2ERC20 won't be able to know about the address of the deposit contract until after it is deployed

File: contracts/ERC20/L2ERC20.sol [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        string memory _tokenName,
+        uint8 _decimalUnits,
+        string memory _tokenSymbol
+    ) public ERC20(_initialAmount, _tokenName, _decimalUnits, _tokenSymbol) {}
+
+    function init(
+        address _messenger,
+        address _L1ERC20DepositAddress
+    ) public {
+        require(l1ERC20DepositAddress == address(0), "L2ERC20 instance has already been initalized");
```
Comment (K-Ho) [inline contracts/ERC20/L2ERC20.sol:23]: I wonder if there could be some potential weirdness if the deployer forgets to `init` and then users try to withdraw or if someone does something like call `init(messenger_address, ZERO_ADDRESS)`? It should be fine though as long as the deployer checks that the messenger and deposit address were `init`ed successfully

File: test/erc20/simplified-message-passing.spec.ts [hunk 13]
```diff
@@ -36,103 +40,126 @@ describe('L1 <=> L2 ERC20 (Simplified Example)', () => {
...
+      L2ERC20.transfer(BobL1Wallet.getAddress(), 2000)
+
+      let alice_l2balance = await L2ERC20.balanceOf(await AliceL1Wallet.getAddress())
+      let alice_l1balance = await L1ERC20.balanceOf(await AliceL1Wallet.getAddress())
+      let bob_l2balance = await L2ERC20.balanceOf(await BobL1Wallet.getAddress())
+      let bob_l1balance = await L1ERC20.balanceOf(await BobL1Wallet.getAddress())
+      assert(alice_l2balance == 3000, `alice l2 balance ${alice_l2balance} != 3000` )
+      assert(alice_l1balance == 5000, `alice l1 balance ${alice_l1balance} != 5000` )
+      assert(bob_l2balance == 2000, `bob l2 balance ${bob_l2balance} != 2000` )
+      assert(bob_l1balance == 0, `bob l1 balance ${bob_l1balance} != 0` )
```
Comment (K-Ho) [inline test/erc20/simplified-message-passing.spec.ts:117]: [nit] using assert here, but expect below (I'd use expect)

File: test/erc20/simplified-message-passing.spec.ts [hunk 14]
```diff
@@ -36,103 +40,126 @@ describe('L1 <=> L2 ERC20 (Simplified Example)', () => {
...
+
+      await L2ERC20.connect(BobL1Wallet).withdraw(1000)
+      await relayL2ToL1Messages(signer)
+
+      alice_l2balance = await L2ERC20.balanceOf(await AliceL1Wallet.getAddress())
```
Comment (K-Ho) [inline test/erc20/simplified-message-passing.spec.ts:122]: [nit] More succinct to replace all of these with single line expects like:
```
expect(
  await L1ERC20.balanceOf(await AliceL1Wallet.getAddress())
).to.be.eq(3000)
```

**Review by K-Ho — APPROVED** [review]
lgtm! Couple of nitpicks to clean up.

 Big thing is checking that (`msg.sender` == the corresponding messenger)  for contracts that receive cross domain messages. I don't think we zero out the `xDomainMessageSender` after a message is relayed so you could have: 
* Alice does deposit, which sets xDomainMessageSender to L1Erc20Deposit and relays `mint` to L2CrossDomainMessenger. * Then Mallory just calls mint - the xDomainMessageSender is still set to the L1Erc20Deposit and Mallory gets infinite internet money

File: contracts/ERC20/L2ERC20.sol (hunk 10 above)
Comment (axchu) [reply contracts/ERC20/L2ERC20.sol:29]: done, plus added a test case for this

File: contracts/ERC20/L1ERC20Deposit.sol (hunk 11 above)
Comment (axchu) [reply contracts/ERC20/L1ERC20Deposit.sol:45]: done, plus added a test case for this

File: contracts/ERC20/L2ERC20.sol (hunk 12 above)
Comment (axchu) [reply contracts/ERC20/L2ERC20.sol:23]: would it be better to do a boolean initialize variable instead? that gets set in the init() so that there are no address checks?
